### PR TITLE
Where cxbyte attempts a comprehensive fix of LibLine, and sprinkles more Utf8Views in it...and ends up refactoring the shit out of it

### PR DESCRIPTION
**Pull request by alimpfard**
- [x] successive tabs after completion should not mess with the anchored styles
- [x] emojis inside suggestions should not mess with the anchored styles
- [x] cycling emoji suggestions should not mess with the anchored styles
- [x] suggestion previews should be spaced properly, even if they have emojis
- [x] anchored styles should not disappear after
    - [x] failed completion attempt
    - [x] successful completion attempt for an unrelated token
- [x] anchored styles should be correctly applied after cycling suggestion pages
- [x] paging should work
- [x] going back a page. (how = by precalculating page boundaries)
- [x] cute little next-page indicator
- [x] redrawing completion previews should not be so dang slow
- [x] Userland/js should not do weird things when suggesting stuff

Please tell me of any edge cases you might think of~

This PR (so far) unbreaks paging because refactoring made that easy.

### CODE REVIEW

**Comment by awesomekling** [comment]
cc @bugaevc

**Comment by alimpfard** [comment]
I always forget to cc him :facepalm:

File: Libraries/LibLine/Editor.cpp [hunk 1]
```diff
@@ -126,6 +126,36 @@ static size_t codepoint_length_in_utf8(u32 codepoint)
     return 3;
 }
 
+Editor::CodepointRange Editor::scan_for_byte_offsets(size_t start_offset, size_t end_offset, size_t scan_offset, bool reverse) const
+{
+    size_t byte_offset = 0;
+    size_t codepoint_offset = scan_offset + reverse;
```
Comment (bugaevc) [inline Libraries/LibLine/Editor.cpp:138]: Adding a `bool` to a `size_t`? If that is what you intended to do, maybe write it explicitly?

File: Libraries/LibLine/Editor.h [hunk 2]
```diff
@@ -293,6 +289,11 @@ class Editor {
     void recalculate_origin();
     void reposition_cursor();
 
+    struct CodepointRange {
+        size_t start { 0 }, end { 0 };
```
Comment (bugaevc) [inline Libraries/LibLine/Editor.h:310]:
```suggestion
        size_t start { 0 };
        size_t end { 0 };
```

Would be more Serenity C++ style

File: Libraries/LibLine/Editor.cpp [hunk 3]
```diff
@@ -126,6 +126,36 @@ static size_t codepoint_length_in_utf8(u32 codepoint)
...
+    CodepointRange offsets;
```
Comment (bugaevc) [inline Libraries/LibLine/Editor.cpp:139]: `start_offset`, `end_offset`, `scan_offset`, `byte_offset`, `codepoint_offset`, `offsets`, oh my!

If you tell me what this method does, I'll think of better names for these all.

File: Libraries/LibLine/Editor.cpp (hunk 1 above)
Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:138]: Fine, I'll cast it.
those C habits die hard

Comment (bugaevc) [reply Libraries/LibLine/Editor.cpp:138]: I mean maybe `reverse ? scan_offset + 1 : scan_offset`

File: Libraries/LibLine/Editor.cpp (hunk 3 above)
Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:139]: This method converts a range described by relative byte offsets ({start,end}_offset) to a codepoint offset (`offsets`), `scan_offset` is where it should start from, and reverse is quite clear.

File: Libraries/LibLine/Editor.cpp (hunk 1 above)
Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp]: but...why introduce a conditional branch that the compiler might decide not to optimise away?

Comment (awesomekling) [reply Libraries/LibLine/Editor.cpp]: Let's always prioritize clarity until profiling proves that something is a bottleneck :)

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp]: Certainly, but `number + boolean` was very clear to me.
regardless, I changed it to be explicit `foo + (bar ? 1 : 0)`

File: Libraries/LibLine/Editor.cpp (hunk 3 above)
Comment (bugaevc) [reply Libraries/LibLine/Editor.cpp:139]: OK, it's taken me some more reads of the function even after your explanation, but I think I get it now.

So the first reason why it's confusing is you call both byte offsets and codepoint indexes *offsets*. Don't, I guess?

What you do is you iterate, over a UTF-32 buffer (`m_buffer`), from `scan_offset` (which is a codepoint index, not a byte offset) either forwards or backwards, and you're counting how many bytes this would take if encoded in UTF-8, without actually encoding it in UTF-8. You're interested in the range of codepoint indexes for a range of byte offsets, `start_offset` and `end_offset`. Is that right?

I would call the method `offset_range_to_codepoint_range()`. I still don't understand why you want to scan in reverse; perhaps instead of scanning from `index_to_scan_from` down to 0 (you're always scanning to the end, don't you?), I'd scan (from 0) up until I find the codepoint index for `end_offset` and stop there. So I wouldn't need the other two arguments. Or I'm misunderstanding how this is supposed to work.

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:139]: my brain just calls relative indices offsets, I can't help it (hence why your help on naming is greatly appreciated!)

on `reverse==true`, you're right, the scan is always from 0.
but otherwise, the scan is not from the beginning or the end of the buffer.
the buffer might be N codepoints long, but the reverse scan might be initiated at codepoint offset M (M < N).

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:139]: Here's a nice diagram:
```
buffer [ 0 1 2 3 . . . A . . . B . . . M . . . N ]
                       ^       ^       ^       ^
                       |       |       |       +- end of buffer
                       |       |       +- scan offset
                       |       +- range end
                       +- range start 
```

Comment (bugaevc) [reply Libraries/LibLine/Editor.cpp:139]: Yes, but instead of M→0 you could be scanning 0→M, right? And then instead of passing M explicitly, you could make the method stop as soon as it hits `end_offset`.

Comment (bugaevc) [reply Libraries/LibLine/Editor.cpp:139]: Nice diagram, perhaps put that, and some of the prose, into the code as comments?

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:139]: How would that work?
I need offsets from M, not from 0

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp:139]: Sure :)

Comment (bugaevc) [reply Libraries/LibLine/Editor.cpp:139]: Ah, you need offsets from M! That wasn't at all clear!

OK, at this point I'd add tons of comments about all of this :smile:

Comment (alimpfard) [reply Libraries/LibLine/Editor.cpp]: If the comments (now) included are not clear enough, please tell me so I can try clarify more~

**Comment by alimpfard** [comment]
This fixes all the issues I could find.
If there is any other issues, _please_ tell me about them.

**Comment by bugaevc** [comment]
Other than that, I still don't understand LibLine internals well enough to tell :|

**Comment by alimpfard** [comment]
\*cough\* sometimes I don't get the thing either, it's become much more complex than I really wanted it to be.

The price to pay for having niceties, I guess.

**Comment by bugaevc** [comment]
No, it's just a sign that you have to rethink the design :wink: Split it into several types, where each one has its own clearly defined scope of responsibility, and doesn't know about internals of others.

(That's a generic advice, I don't have anything specific in mind — in particular because, again, I don't understand it.)

**Comment by alimpfard** [comment]
:thinking: 
my immediate thought is "but they're all interconnected!"
I guess I'll take a few steps back and try to split this thing up.

**Comment by bugaevc** [comment]
Off the top of my head, *presenting* the completions to the user (whether in a terminal that speaks xterm escape codes and positions UTF-8 codepoints on a grid, a terminal that doesn't, or potentially a GUI autocompletion widget, — sounds like an abstract class with a few implementations to me), splitting it into tokens is another thing, and managing completions — finding a common prefix and so on — is yet another thing.

**Comment by alimpfard** [comment]
> presenting the completions to the user [...] potentially a GUI autocompletion widget

I'm not sure I want this library to be used for a GUI widget, it's very much not compatible.

> splitting it into tokens is another thing

Ultimately, I want this line editor to just not care.
I want it to have no such notion of "token", and I am partly there, once I handle this pesky C-w, there will be no concept of "token" in this thing.

> managing completions — finding a common prefix and so on — is yet another thing.

This makes sense to me.

**Comment by bugaevc** [comment]
> I'm not sure I want this library to be used for a GUI widget, it's very much not compatible.

I don't mean using LibLine to drive a `GUI::Model`. I mean having LibLine emit some magical escape sequences that the Terminal will then interpret and use to display suggestions in a GUI popup. (Actually a lot like the recent hyperlinks.)

The point is, you wouldn't have to align the suggestions into columns, you would present them some other way. That means the presenter should not know too much about how the suggestions get generated / decided upon, and the code that generates and filters them should know next to nothing about how they are presented.

**Comment by alimpfard** [comment]
> I mean having LibLine emit some magical escape sequences that the Terminal will then interpret and use to display suggestions in a GUI popup

I love that idea.

**Comment by alimpfard** [comment]
That seemed to be the only reason.
I can't think of any other reason why that specific change would make the whole thing look less jumpy.

Try it, make like 400 files and try to cycle between the suggestions by just holding tab down.
The cursor appears to jump to the middle before that patch, and not after.

**Comment by alimpfard** [comment]
I _did_ confirm that it's not the multiple printfs inside the loop...
Maybe I should try profiling it somehow...

**Comment by alimpfard** [comment]
oh...kay? now it's not jumping around, even with `VT::{save,restore}_cursor` :/
I'll just drop that commit

**Comment by alimpfard** [comment]
The jumping around was because of a `dbg()` in that loop lol

**Comment by bugaevc** [comment]
But why would `dbg()` affect stdout/stderr?

**Comment by alimpfard** [comment]
it probably blocks long enough for the print pause to be visible between the two styling flushes

**Comment by alimpfard** [comment]
@bugaevc @awesomekling, what do yall think about the "more>" at the bottom? should it be a thing or no?

shot-2020-05-22_19-07-23

**Comment by awesomekling** [comment]
I'm kinda liking the `more` thing, although it doesn't make it clear how to actually "get more"

**Comment by alimpfard** [comment]
hmmm...I'd like to keep its footprint to be small, otherwise I'd say "more pages available>"

**Comment by awesomekling** [comment]
> hmmm...I'd like to keep its footprint to be small, otherwise I'd say "more pages available>"

Since you say "pages", perhaps it could say "page 1 of 3" or similar?

**Comment by alimpfard** [comment]
that could work.
`< page 2 / 3 >`...with or without the arrows?

**Comment by alimpfard** [comment]
I'm gonna go with the arrows, I dunno.

**Comment by alimpfard** [comment]
I think that's it for this PR.
if @bugaevc approves, I have nothing to add or change here.

Further splitting of LibLine would be nice, but I don't think I'm in the right headspace for more refactoring work right now.

File: Libraries/LibLine/XtermSuggestionDisplay.cpp [hunk 4]
```diff
@@ -62,8 +64,45 @@ void XtermSuggestionDisplay::display(const SuggestionManager& manager)
         lines_used += max_line_count;
         longest_suggestion_length = 0;
     }
+
     VT::move_absolute(max_line_count + m_origin_x, 1);
 
+    if (m_pages.size() == 0) {
```
Comment (bugaevc) [inline Libraries/LibLine/XtermSuggestionDisplay.cpp:70]: [Please use](https://rust-lang.github.io/rust-clippy/master/#len_zero) `is_empty()` instead of checking for `size() == 0`.

File: Libraries/LibLine/XtermSuggestionDisplay.cpp [hunk 5]
```diff
@@ -62,8 +64,45 @@ void XtermSuggestionDisplay::display(const SuggestionManager& manager)
...
+        size_t num_printed = 0;
+        size_t lines_used { 1 };
```
Comment (bugaevc) [inline Libraries/LibLine/XtermSuggestionDisplay.cpp:72]:
```suggestion
        size_t lines_used = 1;
```

for trivially constructed locals is the Serenity style.

Comment (alimpfard) [reply Libraries/LibLine/XtermSuggestionDisplay.cpp]: Noted.

File: Libraries/LibLine/XtermSuggestionDisplay.cpp [hunk 6]
```diff
@@ -31,7 +31,7 @@
 
 namespace Line {
 
-void DumbTerminalSuggestionDisplay::display(const SuggestionManager& manager)
+void XtermSuggestionDisplay::display(const SuggestionManager& manager)
```
Comment (bugaevc) [inline Libraries/LibLine/XtermSuggestionDisplay.cpp:36]: This doesn't seem to belong in this commit.

File: Libraries/LibLine/XtermSuggestionDisplay.cpp (hunk 4 above)
Comment (alimpfard) [reply Libraries/LibLine/XtermSuggestionDisplay.cpp]: Does this really make sense in this context? we know that `Vector::size()` is just a member access, inline at that...

Applied nonetheless.

File: Libraries/LibLine/XtermSuggestionDisplay.cpp (hunk 6 above)
Comment (alimpfard) [reply Libraries/LibLine/XtermSuggestionDisplay.cpp:36]: I thought I moved that change to its origin commit...ouch.

Comment (alimpfard) [reply Libraries/LibLine/XtermSuggestionDisplay.cpp:36]: Fixing that is gonna _hurt_...T_T

Comment (alimpfard) [reply Libraries/LibLine/XtermSuggestionDisplay.cpp:36]: Fixed :)

File: Libraries/LibLine/SuggestionManager.cpp [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+            result.offset_region_to_remove = { m_next_suggestion_invariant_offset, shown_length };
+            result.new_cursor_offset = actual_offset;
+        }
+
+        auto& suggestion = suggest();
+        set_current_suggestion_initiation_index(initiation_start_index);
+
+        if (mode == CompletePrefix) {
+            // only auto-complete *if possible*
```
Comment (awesomekling) [inline Libraries/LibLine/SuggestionManager.cpp:151]: https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments

Comment (alimpfard) [reply Libraries/LibLine/SuggestionManager.cpp:151]: uhhh...capitalisation?

Comment (alimpfard) [reply Libraries/LibLine/SuggestionManager.cpp:151]: Mind if I do this in a separate commit?
tracking the comments through the history is a bit...annoying.
I could do that if needed though.

Comment (awesomekling) [reply Libraries/LibLine/SuggestionManager.cpp:151]: That's fine! :)

Comment (alimpfard) [reply Libraries/LibLine/SuggestionManager.cpp:151]: Github seems very broken today...

anyway, proper comments go _brrrrrrrrrrrrrrrrrrrrrrrrrrrrrr_

**Comment by awesomekling** [comment]
:star:
